File: core/src/commonMain/kotlin/work/socialhub/kmastodon/api/TimelinesResource.java
package work.socialhub.kmastodon.api;

import mastodon4j.Range;
import mastodon4j.entity.Conversation;
import mastodon4j.entity.Status;
import mastodon4j.entity.share.Response;

/**
 * @author hecateball
 */
public interface TimelinesResource {

    /**
     * Retrieving a home timeline.
     *
     * @return an array of Statuses, most recent ones first
     */
    public Response<Status[]> getHomeTimeline(
            Range range);

    /**
     * Retrieving a public timeline.
     *
     * @param local (optional) only return statuses originating from this instance
     * @return an array of Statuses, most recent ones first
     */
    public Response<Status[]> getPublicTimeline(
            Boolean local,
            Boolean onlyMedia,
            Range range);

    /**
     * Retrieving a tag timeline.
     *
     * @param hashtag
     * @param local   (optional) only return statuses originating from this instance
     * @return an array of Statuses, most recent ones first
     */
    public Response<Status[]> getHashtagTimeline(
            String hashtag,
            Boolean local,
            Boolean onlyMedia,
            Range range);

    /**
     * Retrieving a list timeline.
     *
     * @param listId
     * @return an array of Statuses, most recent ones first
     */
    public Response<Status[]> getListTimeline(
            String listId,
            Range range);

    /**
     * Retrieving a conversations.
     *
     * @return an array of Conversations.
     */
    public Response<Conversation[]> getConversations(
            Range range);
}

File: core/src/commonMain/kotlin/work/socialhub/kmastodon/api/TimelinesResource.kt
package work.socialhub.kmastodon.api

import work.socialhub.kmastodon.api.request.timelines.TimelinesConversationsRequest
import work.socialhub.kmastodon.api.request.timelines.TimelinesHashTagTimelineRequest
import work.socialhub.kmastodon.api.request.timelines.TimelinesHomeTimelineRequest
import work.socialhub.kmastodon.api.request.timelines.TimelinesListTimelineRequest
import work.socialhub.kmastodon.api.request.timelines.TimelinesPublicTimelineRequest
import work.socialhub.kmastodon.api.response.Response
import work.socialhub.kmastodon.api.response.timelines.TimelinesConversationsResponse
import work.socialhub.kmastodon.api.response.timelines.TimelinesHashTagTimelineResponse
import work.socialhub.kmastodon.api.response.timelines.TimelinesHomeTimelineResponse
import work.socialhub.kmastodon.api.response.timelines.TimelinesListTimelineResponse
import work.socialhub.kmastodon.api.response.timelines.TimelinesPublicTimelineResponse
import kotlin.js.JsExport

@JsExport
interface TimelinesResource {

    /**
     * Retrieving a home timeline.
     */
    suspend fun homeTimeline(
        request: TimelinesHomeTimelineRequest
    ): Response<Array<TimelinesHomeTimelineResponse>>

    @JsExport.Ignore
    fun homeTimelineBlocking(
        request: TimelinesHomeTimelineRequest
    ): Response<Array<TimelinesHomeTimelineResponse>>

    /**
     * Retrieving a public timeline.
     */
    suspend fun publicTimeline(
        request: TimelinesPublicTimelineRequest
    ): Response<Array<TimelinesPublicTimelineResponse>>

    @JsExport.Ignore
    fun publicTimelineBlocking(
        request: TimelinesPublicTimelineRequest
    ): Response<Array<TimelinesPublicTimelineResponse>>

    /**
     * Retrieving a tag timeline.
     */
    suspend fun hashtagTimeline(
        request: TimelinesHashTagTimelineRequest
    ): Response<Array<TimelinesHashTagTimelineResponse>>

    @JsExport.Ignore
    fun hashtagTimelineBlocking(
        request: TimelinesHashTagTimelineRequest
    ): Response<Array<TimelinesHashTagTimelineResponse>>

    /**
     * Retrieving a list timeline.
     */
    suspend fun listTimeline(
        request: TimelinesListTimelineRequest
    ): Response<Array<TimelinesListTimelineResponse>>

    @JsExport.Ignore
    fun listTimelineBlocking(
        request: TimelinesListTimelineRequest
    ): Response<Array<TimelinesListTimelineResponse>>

    /**
     * Retrieving a conversations.
     */
    suspend fun conversations(
        request: TimelinesConversationsRequest
    ): Response<Array<TimelinesConversationsResponse>>

    @JsExport.Ignore
    fun conversationsBlocking(
        request: TimelinesConversationsRequest
    ): Response<Array<TimelinesConversationsResponse>>
}
